lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs

```diff
@@ -1206,15 +1210,21 @@ prop_balanceTxValid
                     ">20 payment outputs"
                 . classify (length originalOuts > 100)
                     ">100 payment outputs"
-
+        let coverage res =
+                cover 10 (isRight res) "success"
+                -- We're not using 'checkCoverage', as we like to keep direct
+                -- control over the number of tests we run. We are also
+                -- intentionally keeping the "success" /label/ below for the
+                -- sake of preserving the pattern of (at least) one label per
+                -- case branch.
```

Comment:
Hi @Anviking!

I'm not sure I understand the motivation here.

On the one hand, the code seems to be saying that we do want at least 10% coverage of the "success" case (hence the call to `cover 10`).

But on the other hand, the lack of `checkCoverage` means that the test will not fail if the coverage condition is not reached. The comment seems to imply that it's more important to control the number of tests we run.

Why is it so important to control the exact number of tests?

The disadvantage of this approach is that it's possible for a future commit to change the implementation (or generators) in such a way that the "success" condition is never reached, and the tests will still pass.

Reply:
> Why is it so important to control the exact number of tests?

Two things that matter are 1) the time it takes to run the test 2) the number (not ratio) of each interesting case we hit.

What I don't see mattering are the exact `cover x` targets, nor the extreme confidence `checkCoverage` normally seeks:

> If the coverage requirement is met, and the certainty parameter is n, then you should get a false positive at most one in n runs of QuickCheck. The default value is 10^9.

What if the success rate becomes unexpectedly big? It seems we could end up with just 100 tests:

```
    produces valid transactions or fails (Babbage) [✔] (2440ms)
      +++ OK, passed 100 tests:
      51% success
    produces valid transactions or fails (Conway) [✔] (2854ms)
      +++ OK, passed 100 tests:
      51% success
```

effectively running the tested successes from 110 (1000 * 0,11) to 51 (0,51 * 100) — which, fair enough, is more acceptable than accidentally going from 1000*0,11 to 1000*0 or 1000*0,01.

So I have now added the `checkCoverage`, but I wish there was a better way 😶

Reply:
> 1) the time it takes to run the test

Are you concerned that the test suite will become too slow?

> 2) the number (not ratio) of each interesting case we hit.

Perhaps the real problem here is that we're placing too many expectations on this property test. Currently, we're trying to use a single generator to cover:
 - all the possible failure cases
 - all of the interesting success cases
 
 This means that any increase in coverage for one interesting case puts us at risk of accidentally decreasing the coverage of some other interesting case.
 
We've also using `withMaxSuccess 1000`, which is presumably an attempt to make sure that all of the individual interesting cases are each covered with at least a few test runs.

But instead of using the above approach, perhaps it would be better to split up this "mega-property" into multiple smaller properties, where each property test:
- has its own notion of expected coverage;
- can use a generator of choice, tailored to meet the coverage condition.

Reply:
We could presumably split the property in two, one with a `isLeft res ==> prop` precondition, and one with `isRight res ==> prop` (or I think you were advocating multiple properties for the isLeft branch). This would make coverage easier to control, but it would waste CPU resources on generating test cases just to discard them. Though for something as notable as Right vs Left, it might be worth it.

> Are you concerned that the test suite will become too slow?

In the sense that I always want to ensure we are running as many tests as possible with the constraint of them passing in a reasonable time, yes.

IIRC the ledger team did at some point move from "generate envs for each test" to "run each test for every env generated" and were very happy about it.